README.md
#SelfOrganizingMap


|Since		|09/04/14		|
|----------:|--------------:|
|Written by |Daniel Swain Jr|

##An alteration to the original SelfOrganizingMap.

#SetCreator  		

Creates the test and trainingSet from the input file
using the configurations in the config file.
TestSets are created by randomly grabbing 10% of 
each record of each decision 10 times. TrainingSets
are then created by taking the leftover 90% of each of the TestSets. 
TrainingSets are then trained by sending records into
an matrix that begins empty. When sent, 1 of 3 cases can happen:

* If the record sent is a neighbor to any of the rules
 in the matrix, the record updates the rule using an 
 algorithm and then added to the cluster of said rule.

* If the record is a neighbor to more than one rule, the decision
 of the record is used to decide which rule is updated by the 
 record. If they match, the rule is updated by the record.

* If the record is a neighbor to none of the rules, then 
 the record itself becomes a rule and is appended to the end
 of the matrix.

This is how trainingSets are created.

##Intersector

Takes the testSets and sends them to the corresponding 
trained trainingSet.


changelog.md
#SelfOrganizingMapV2  
##Change Log

##Written by Daniel Swain Jr

###1.0 	- 09/04/14

	+ The project works, but its untidy. Fix iminent

###1.0.1 	- 09/09/14

	+ Added some functionalities.

###1.1 	- 09/16/14

	+ Complete overhaul of the structure of the code.
	+ Much easier to read and stuff.

###1.1.1 	- 09/23/14

	+ Added more functionalities. 

###1.1.2 	- 09/30/14

	+ Fixed: RandRecordSet
		- getLeftovers() 	
			was returning completely wrong outputs. FIXED
		- createTrainingSet	
			because getLeftovers() wasn't working,
			this wasn't either. FIXED

	+ Made SOMEnhanced better:
		- firstTimeInitialization() is now run()
		- aquireTrainingSet() is now train(int numOfSets)
		- now accurately prints 10 trainingSets with readable records

	+ Known Bugs:
		- Need to make the output and trainingSet file numbers the 
			same in every situation so it's easier to find the pairs
		- When making more than one training set, the records start 
			to come out funny. Training sets 1 and 2 look decent.

###1.1.3 	- 10/02/14

	+ Fixed the mysterious winner rule
		- cleanMatrix()		
			Added a feature so that it will now clean the winners
			 that don't have corresponding records in the matrix

	+ Changed the name of:
		- SelfOrganizingMap =>	AbstractSetCreator
		- SOM  				=>	SetCreator
		- SOMEnhanced		=>	SetCreatorEnhanced
		- Should be easier to understand now.

	+ TODO:
		- Build new package containing classes to handle the
			shooting of the test sets to the chosen training set
		
###1.1.4	- 10/05/14

	+ Trying to fix: Multiple TrainingSet Problem
		- Currently when making multiple training sets, the records
		 	change throughout the queues, i.e. the records altered in 
		 	the previous run are the same records used over again.

###1.1.5	- 10/05/14

	+ Kind of fixed: Multiple TrainingSet Problem
		- Creating multiple trained sets used to be a problem,
			 and is now fixed by the addition of the resetAlpha() function
		- Alpha wasn't being reset, therefore the updated values
			 became smaller and smaller

	+ New Problem: Duplicates... Again.
		- Duplicates are showing up again even though being ran 
			through cleanMatrix()

###1.2.0 	- 10/07/14

	+ Kind of fixed: Certainty Factor/Cover Problem
		- If a Rule(aka Seed) had it's dominant decision changed
			it wouldn't count itself as a winner and the cover
			and certainty factor would be off.

	+ Added functions to: Intersector
		- (void)read(int) 	
			// Reads the files of the corresponding int
		- (void)readTestSetFile(int)	
			// Reads the test set file of the corresponding int
		- (void)readTrainingSetFile(int)
			// Reads the training set file of the corresponding int
		- (void)run()		
			// Runs the algorithm

	+ Added functions to: Record
		-Record(Record,String)
			// Constructor that creates a record with a name

	+ Still A Problem: Duplicates... Again.
		- The duplicates seem to be somewhat different from each other
			therefore are not being caught by the cleanMatrix()
		- Will ask Dr Hashemi about this.

	+ TODO:
		- implement run() in Intersector
		- Do something about the semi-duplicates

###1.2.1 	- 10/08/14
	
	+ Created:
		- Intersector.runAlgorithm() - Sends each test record against
			all rules in the trainingSet
		- Intersector.sumAbsDif() - Returns the sum of the absolute 
			value differences ofattributes between two given records.
		- Intersector.times	- Number of files/times to do the algorthim
		- Intersector.match - A map where the key is the file number
			and the value is a double array of size 2 containing:
				1. The # of Matches that came through
				2. The Size of the TrainingSet

	+ Implemented run()
		-Uses runAlgorithm on each of the specified number of sets

	+ Changed the name of:
		- Record.isWinner() => Record.hasSameDecision()

	+ TODO:
		- Make into files
		- Other things...

###1.2.2 	- 10/08/14

	+ Created:
		-Intersector.results() - ArrayList<String> that contains the formatted
			results of each run.
		-Intersector.resultsToFile(String) - Creates files in a directory
			for the results

	+ Changes in:
		- Intersector.runAlgorithm() - Added functionalities for printing

	+ TODO:
		- The theres an offset because printing of the files starts at
			1, not a huge problem, but should go back and fix.

###1.2.3 	- 10/09/14

	+Created:
		-Intersector.totalResultsFile() - takes the totals of all tests and
			puts them in one file called "overall.txt"

	+ TODO:
		- Round off the records in the training sets to the thousandths 
			after training is complete and re-clean matrix

###1.2.4	- 10/13/14

	+ Created:
		- Matrix.round(double, int) - rounds the given value to the given number of places

	+Changes in:
		- Intersector.sumAbsDif() - changes so that the values added are either 0 or 1.
		- Intersector.runAlgorithm() - added functionality for true/false pos/neg table.
		- Matrix.updateRecord() - values are now rounded to the thousandths

	+ TODO:
		- Negative values... How are they made?!

###2.0.01	- 10/16/14
	
	Summary:
		The SetCreatorEnhanced.shootRecords() has been changed so that in the case of 
		multiple rules being the neighbor of a record and also sharing the same decision,
		the node with the largest cluster is chosen.

	+ Changed the name of:
		- SetCreatorEnhanced.findWinner() => doAlgorithm()
		- MatrixEnhanced.winners => clusters

	+ Created:
		- SetCreator.readOrigTrainingSetFile() - reads in the origTrainingSet files from
		the trainingSets folder and adds them to the trainingSet array.
		- SetCreatorEnhanced.runForPreCreatedSets() - does the same as run except that
		it's for pre-created training sets
		- SetCreatorEnhanced.trainPreCreatedSets() - trains all pre-created training sets
		- SetCreatorEnhanced.shootRecords_2() - does the same as shootRecords except that
		it's for pre-created training sets and does not need to use RandomSet.createQueue()
		and also does not print the origTrainingSets
		- MatrixEnhanced.getClusterSize() - returns the size of a rule's cluster

	+ Changes in:

src/com/wiseaux/setCreator/SetCreatorEnhanced.java
package com.wiseaux.setCreator;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Random;

/**
 * This class is responsible for sending records into the given matrix
 *
 * @author Daniel Swain
 * @version 2.0.01
 * @since 9/04/2014
 */
public class SetCreatorEnhanced extends SetCreator {

    /**
     * Debugging strings.
     */
    ArrayList<String> debug = new ArrayList<>();
    String table = "%-10s|%-10s|%-10s|%-10s\n";
    String line = "----------|----------|----------|----------\n";
    /**
     * Handle creation of the sets
     */
    private RandRecordSet rSet;

    public SetCreatorEnhanced() {
    }

    /**
     * Reads the configuration file, initializes the records array, reads the
     * input file, initializes rSet, creates the testSets, creates the
     * trainingSet and initializes seeds.
     */
    public void run() {
        readConfig();
        initializeAllRecords(numOfRecords, numOfAttributes);
        readRecords();

        initDebug();//Debug

        rSet = new RandRecordSet(allRecords);
        testSet = rSet.createTestSet(allRecords, 0, 1);
        trainingSet = rSet.createTrainingSet(testSet, 0.1); // using 90%
        matrices = new MatrixEnhanced[trainingSet.length];

        testSetsToFiles();
    }

    /**
     * Reads the configuration and trains with the pre-created sets
     */
    public void runForPreCreatedSets() {
        readConfig();
        rSet = new RandRecordSet();
    }

    /**
     * Reads the configuration and trains k-1 records
     */
    public void runForSpecialSet() {
        readConfig();
        initializeAllRecords(numOfRecords, numOfAttributes);
        readRecords();
        rSet = new RandRecordSet(allRecords);
        testSet = rSet.createSpecialTestSet(allRecords, 0, 1);
        trainingSet = rSet.createSpecialTrainingSet(testSet, 0.01); // using 99%
        matrices = new MatrixEnhanced[trainingSet.length];

        specialTestSetsToFiles();
    }

    /**
     * Puts the testSets into files
     */
    public void testSetsToFiles() {
        Queue[] temp = new Queue[testSet.length];
        for (int i = 0; i < temp.length / 2; i++) {
            temp[i] = rSet.createQueue(testSet, i, testSet.length);
        }
        testSet = temp;
        new File("testSets").mkdir();

        for (int i = 0; i < testSet.length / 2; i++) {
            rSet.printQue(testSet[i], "testSets\\testSet");
        }
    }

    /**
     * Puts the specialTestSets into files
     */
    public void specialTestSetsToFiles() {
        Queue<Record> tmp = new LinkedList<>();

        for (Queue<Record> test : testSet) {
            tmp.addAll(test);
        }
        rSet.setNamesForRecords(tmp);

        new File("testSets").mkdir();

        rSet.printQue(tmp, "testSets\\specialTestSet");

    }

    /**
     * Access shootRecords() function in order to train the trainingSet
     */
    public void train(int numOfSets) {
        new File("trainingSets").mkdir();
        shootRecords(trainingSet, numOfSets, matrices);

        //debuggingFile();

    }

    /**
     * Access shootRecordsForPreCreatedSets() using the pre-created
     * trainingSets.
     */
    public void trainPreCreatedSets() {
        trainingSet = new Queue[10];
        matrices = new MatrixEnhanced[trainingSet.length];
        fillTrainingSet();
        shootRecords_2(trainingSet, trainingSet.length, matrices);
        this.outputToFile(matrices, "trainingSets\\trainingSetWithData");
    }

    /**
     * Access shootRecords() using a set of k-1 records
     */
    public void trainAllButOneRecord() {
        new File("trainingSets").mkdir();
        shootRecords_3(trainingSet, trainingSet.length / 2);
    }

    /**
     * Creates a queue to send to doAlgorithm using the given queue array
     */
    private void shootRecords(Queue[] records, int numOfSets, MatrixEnhanced[] matrices) {
        for (int i = 0; i < numOfSets; i++) {
            Queue<Record> que = rSet.createQueue(records, i, records.length);

            MatrixEnhanced matrix = new MatrixEnhanced();
            resetAlpha();

            if (que.isEmpty()) {
                System.out.println("There are no records to send!");
                System.exit(1);
            }
            rSet.printQue(que, "trainingSets\\origTrainingSet");
            doAlgorithm(que, epochs, matrix);
            matrix.combineRegions();
            matrix.checkForDuplicates();
            matrix.addCertainties();
            matrix.printForRead();
            this.outputToFile(matrix, "trainingSets\\trainingSetWithData");
        }
    }

    /**
     * Creates a queue to send to doAlgorithm using the given queue array; for
     * pre-created trainingSets
     */
    private void shootRecords_2(Queue[] records, int numOfSets, MatrixEnhanced[] matrices) {
        for (int i = 0; i < numOfSets; i++) {
            Queue<Record> que = records[i];

            matrices[i] = new MatrixEnhanced();
            resetAlpha();

            if (que.isEmpty()) {
                System.out.println("There are no records to send!");
                System.exit(1);
            }
            //rSet.printQue(que, "trainingSets\\origTrainingSet");
            doAlgorithm(que, epochs, matrices[i]);
            matrices[i].combineRegions();
            matrices[i].checkForDuplicates();
            matrices[i].addCertainties();
            matrices[i].printForRead();
        }
    }

    /**
     * Creates a queue to send to doAlgorithm using the given queue array
     */
    private void shootRecords_3(Queue[] records, int numOfSets) {
        for (int i = 0; i < numOfSets; i++) {
            Queue<Record> que = rSet.createQueue(records, i, records.length);

            MatrixEnhanced matrix = new MatrixEnhanced();
            resetAlpha();

            if (que.isEmpty()) {
                System.out.println("There are no records to send!");
                System.exit(1);
            }
            rSet.printQue(que, "trainingSets\\origTrainingSet");
            doAlgorithm(que, epochs, matrix);
            matrix.combineRegions();
            matrix.checkForDuplicates();
            matrix.addCertainties();
            matrix.printForRead();
            this.outputToFile(matrix, "trainingSets\\trainingSetWithData");
        }
    }

    /**
     * Takes a queue of records and shoots them at the seeds, checking if they
     * are neighbors with each other. If so, then add them to winners and update
     * the matrix.
     *
     * @see isNeighbor
     */
    private void doAlgorithm(Queue<Record> que, int epochCount, MatrixEnhanced matrix) {
        int epoch = epochCount;
        //Debug
        int nonMatchCase = 0;
        int matchCase = 0;

        if (que.isEmpty()) {
            System.out.print("Que is empty");
            System.exit(1);
        }

        //Initialize newRegion
        Map<Integer, HashSet<Record>> newRegion = new HashMap<>();
        for (int i = 0; i < 1000; i++) {
            newRegion.put(i, new HashSet<Record>());
        }

        Queue<Record> inputQueue = new LinkedList<>(que);
        boolean[] removeFromQue = new boolean[que.size()];
        boolean recordRemoved = false;

        int recNum = 1;
        //for each inputRecord in the queue
        while (!inputQueue.isEmpty()) {
            Record inputRecord = inputQueue.remove();
            //Offset neighbors by 1 to include first epoch when record_1 becomes seed_1            
            boolean[] neighbors = new boolean[matrix.getSize() + 1];

            //Used to count the amount of neighbors between the inputRecord and seeds
            int count = 0;

            //Reason for <= : Check the 0th value as well
            for (int i = 0; i <= matrix.getSize(); i++) {
                if (inputRecord.isNeighbor(matrix.getRecord(i), delta)) {
                    neighbors[i] = true;
                    count++;
                }
            }

            if (count == 0) {   //No matches, add to matrix, add to winner, mark to remove
                matrix.addRecord(inputRecord);
                matrix.addWinner(inputRecord, inputRecord.getIntName()); //Update newRegion
                newRegion.get(inputRecord.getIntName()).add(inputRecord);

                removeFromQue[recNum - 1] = true;
                recordRemoved = true;
            } else if (count == 1) { //Find which inputRecord it matched, update
                int index = findTrue(neighbors);
                int name = matrix.getRecord(index).getIntName();
                matrix.updateMatrix(inputRecord, index, alpha, name); // Update newRegion
                newRegion.get(name).add(inputRecord);
                matrix.checkDominant(index, newRegion.get(name));
            } else if (count > 1) { //Find which inputRecord it matched and matched decision, update

                //debug.add("\n\nCase 3\nCount : " + count);
                int nonMatches = 0;//Debug
                int matches = 0;//Debug

                int matchMax = Integer.MIN_VALUE;
                int nonMatchMax = Integer.MAX_VALUE;
                Map<Integer, Record> competitors = new HashMap<>();
                for (int index = 0; index < neighbors.length; index++) {
                    if (neighbors[index]
                            && matrix.getRecord(index).hasSameDecision(inputRecord)) {

                        matches++;//Debug

                        int temp = newRegion.get(index).size();
                        if (matchMax < temp) {
                            competitors.clear();
                            competitors.put(index, matrix.getRecord(index));
                        } else if (matchMax == temp) {
                            competitors.put(index, matrix.getRecord(index));
                        }
                    } else if (neighbors[index]
                            && !(matrix.getRecord(index).hasSameDecision(inputRecord))
                            && (matchMax == Integer.MIN_VALUE)) {

                        nonMatches++;//Debug

                        int temp = newRegion.get(index).size();
                        if (nonMatchMax > temp) {
                            competitors.clear();
                            competitors.put(index, matrix.getRecord(index));
                        } else if (nonMatchMax == temp) {
                            competitors.put(index, matrix.getRecord(index));
                        }
                    }
                }

                //Debug
                if (matches > 0) {
                    matchCase++;
                    //debug.add("\nMatching decision found : " + matches);
                } else {
                    nonMatchCase++;
                    //debug.add("\nNo matching decision found : " + nonMatches);
                }

                if (!competitors.isEmpty()) {
                    Random random = new Random();
                    List<Integer> keys = new ArrayList<>(competitors.keySet());
                    int randomKey = keys.get(random.nextInt(keys.size()));
                    Record winner = competitors.get(randomKey);
                    int name = winner.getIntName();
                    // Update
                    matrix.updateMatrix(inputRecord, randomKey, alpha, name);
                    newRegion.get(name).add(inputRecord);
                    
                    matrix.checkDominant(randomKey, newRegion.get(name));
                    
                }
            }
            recNum++;
        }
        
        debug.add(String.format(table, epoch,
                                       matchCase,
                                       nonMatchCase,
                                       (matchCase + nonMatchCase)
        ));

        matrix.addToRegions(newRegion);
        // If the epoch is at an interval of 10% of the original epoch
        // the update alpha
        if (epoch % (epoch * 0.1) == 0) {
            updateAlpha();
        }

        epoch -= 1;

        if (recordRemoved && epoch != 0) {
            doAlgorithm(rSet.removeRecords(que, removeFromQue), epoch, matrix);
        } else if (epoch != 0) {
            doAlgorithm(que, epoch, matrix);
        }
        
        if(epoch == 0) {
            debug.add("\n");
            initDebug();
        }
    }

    /**
     * Finds the true value in a boolean array
     */
    private static int findTrue(boolean[] boolArray) {
        int index = -1;
        for (int i = 0; i < boolArray.length; i++) {
            if (boolArray[i]) {
                index = i;
            }
        }

        if (index < 0) {
            return index;
        }
        return index;
    }

    private void initDebug() {
        debug.add(String.format(table, "Epoch","Match","NonMatch","Total"));
        debug.add(line);
    }

    private void debuggingFile() {
        String name = "debug";
        File output = new File(name + 1 + ".txt");
        try {
            int i = 1;
            while (output.exists()) {
                output = new File(name + i++ + ".txt");
            }
            output.createNewFile();

        } catch (IOException ex) {
            System.out.print("Can't find directory");
            System.exit(1);
        }

        try (PrintWriter write = new PrintWriter(output)) {

            write.flush();
            for (String str : debug) {
                write.write(str);
            }
            write.close();

        } catch (FileNotFoundException e) {
            System.out.println("Output file not found.");
            System.exit(1);
        }
    }

}
